### DISCUSSION TITLE
Welcome to HomebridgeMagicHome-DynamicPlatform Discussions!

### DISCUSSION THREAD

**Discussion by Zacknetic** [Zacknetic/HomebridgeMagicHome-DynamicPlatform#67] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by EzzatQ** [comment]
Hey! Thanks for this awesome Plugin!

I have a dozen magic home controllers (RGB, GRB, GRBWW) but all my light strips are RGB.
Therefore for some of my light strips (the ones connected to a GRB  or GRBWW controller) the red and green values are inverted. I've looked through the code to try and see if for these specific controllers (ex: GRB), I can set them up as RGB so that the colors won't be inverted.
Could you point me to where in the code I can do something like this?
(or even better, give us an option to set the controller/light the in the config)

Thanks in advance

**Reply by Zacknetic** [reply]
Could you please tell me which controller type it is? A link to amazon would best. Additionally, could you please upload your log file. That way I can see which device type it is. 

On some controllers, you can permanently change the color order by using a IR remote. This is because some LED strips have their colors in a different order. That could also be what is causing the issue.

If the above solutions don't fix your issue I'd be happy to point you in the direction of where the code can be altered. But I have no plans right now to implement a feature allowing users to choose color order. This is because the goal of this plugin is to require as little setup as possible. 

Singling out individual controllers would not stay consistent to that vision. This plugin is always open to pull requests if that is something you'd have the ability and motivation to do. The feature would have to fall under certain parameters which we could discuss.

Let me know how that goes.

**Reply by EzzatQ** [reply]
I have a few different ones but they all resemble this one:
https://www.amazon.com/dp/B075YS3THJ/ref=cm_sw_r_cp_api_glt_fabc_XZS2ZD54XBS32NY6231K
but none of them have an IR receiver (I did have some with IR Receivers but the leds would change every time I controlled my TV or any other device that was controlled by IR)

I did continue to read the code. I see that there's a LightMap file, so that's where ill choose which version I want for which controller, but I can't find the place where you construct a new accessory and set its attributes

**Reply by EzzatQ** [reply]
and here are the logs: 
`[2/23/2021, 9:27:49 PM] [homebridge-magichome-dynamic-platform] 1 - Registering existing accessory...
Display Name: 'Side' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ200' 
Unique ID: 'DC4F228D81B5' 
IP-Address: '10.0.0.14' 
Hardware Version: '44' 
Firmware Version: '8' 

[2/23/2021, 9:27:49 PM] [homebridge-magichome-dynamic-platform] 2 - Registering existing accessory...
Display Name: 'Night Light' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ210' 
Unique ID: 'C44F338FCB55' 
IP-Address: '10.0.0.15' 
Hardware Version: '44' 
Firmware Version: '9' 

[2/23/2021, 9:27:50 PM] [homebridge-magichome-dynamic-platform] 3 - Registering existing accessory...
Display Name: 'Stairs 2' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ210' 
Unique ID: '840D8E69E47A' 
IP-Address: '10.0.0.13' 
Hardware Version: '44' 
Firmware Version: '9' 

[2/23/2021, 9:27:50 PM] [homebridge-magichome-dynamic-platform] 4 - Registering existing accessory...
Display Name: 'Bar' 
Controller Logic Type: 'RGBWStrip'  
Model: 'AK001-ZJ200' 
Unique ID: 'C44F33A15115' 
IP-Address: '10.0.0.23' 
Hardware Version: '4' 
Firmware Version: '5' 

[2/23/2021, 9:27:50 PM] [homebridge-magichome-dynamic-platform] 5 - Registering existing accessory...
Display Name: 'Bedroom Main' 
Controller Logic Type: 'RGBWWBulb'  
Model: 'AK001-ZJ2101' 
Unique ID: '2CF432CE8C3C' 
IP-Address: '10.0.0.10' 
Hardware Version: '35' 
Firmware Version: '7' 

[2/23/2021, 9:27:51 PM] [homebridge-magichome-dynamic-platform] 6 - Registering existing accessory...
Display Name: 'Living Room Main' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ210' 
Unique ID: 'C44F338F74CC' 
IP-Address: '10.0.0.11' 
Hardware Version: '44' 
Firmware Version: '9' 

[2/23/2021, 9:27:51 PM] [homebridge-magichome-dynamic-platform] 7 - Registering existing accessory...
Display Name: 'Tv' 
Controller Logic Type: 'GRBStrip'  
Model: 'AK001-ZJ210' 
Unique ID: 'D8F15B844895' 
IP-Address: '10.0.0.9' 
Hardware Version: '33' 
Firmware Version: '6' 

[2/23/2021, 9:27:51 PM] [homebridge-magichome-dynamic-platform] 8 - Registering existing accessory...
Display Name: 'Strip A37D11' 
Controller Logic Type: 'RGBWStrip'  
Model: 'AK001-ZJ200' 
Unique ID: 'C44F33A37D11' 
IP-Address: '10.0.0.81' 
Hardware Version: '4' 
Firmware Version: '5' 

[2/23/2021, 9:27:52 PM] [homebridge-magichome-dynamic-platform] 9 - Registering existing accessory...
Display Name: 'Ceiling' 
Controller Logic Type: 'RGBWStrip'  
Model: 'AK001-ZJ200' 
Unique ID: 'C44F33A15066' 
IP-Address: '10.0.0.8' 
Hardware Version: '4' 
Firmware Version: '5' 

[2/23/2021, 9:27:52 PM] [homebridge-magichome-dynamic-platform] 10 - Registering existing accessory...
Display Name: 'Study' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ210' 
Unique ID: 'C44F338FE66D' 
IP-Address: '10.0.0.16' 
Hardware Version: '44' 
Firmware Version: '9' 

[2/23/2021, 9:27:52 PM] [homebridge-magichome-dynamic-platform] 11 - Registering existing accessory...
Display Name: 'Simple GRB' 
Controller Logic Type: 'RGBWBulb'  
Model: 'AK001-ZJ200' 
Unique ID: 'DC4F22F026B9' 
IP-Address: '10.0.0.12' 
Hardware Version: '44' 
Firmware Version: '8' 
`

**Reply by EzzatQ** [reply]
Btw should I open an issue to continue this discussion or is it ok to do so here?

**Reply by Zacknetic** [reply]
If you'd like, you're free to open an issue tagged as a feature request. 

The line of code will be in the 'accessories' folder in each accessory class. Find a similar line in each accessory type but keep in mind, doing so will change the byte order for all devices of that type.

Take a look here: 
https://github.com/Zacknetic/HomebridgeMagicHome-DynamicPlatform/blob/master/src/accessories/RGBWWStrip.ts#L76

Let me know if you'd like to do some work towards a more granular solution or if this suits your needs.

**Reply by EzzatQ** [reply]
This would fix the color issue, however I failed to mention that some controllers are RGBWW/GRBWW while the light strips are only RGB and so when I turn the lights on to white, the controller switches to white mode and the colors turn off completely.
What would suit my issue better is probably to manually set the controller's type. (for example, have mu GRBWW controller register as a RGB controller

**Reply by Zacknetic** [reply]
Ah, okay I understand the root of the problem now. 

This can be solved two ways: 

- Changing the config settings for simultaneous device saturation threshold. That way the white never turns on. (and RGB never turns off)
- Adding a feature which allows users to specify the controller type in the HomeKit name (my plugin can read names from HK. i.e. "Window Light $RGB" would set the device to RGB mode and erase $RGB from the name)

**Reply by EzzatQ** [reply]
Ok so I tried to change the code you suggested and the saturation threshold and it worked beautifully! Thank you very much!

I actually would be interested in working on a feature like this but I'm not really familiar with how to implement it with home bridge.
Would be nice to learn though (once I finish my exams) 😅

Again thank you very much

**Reply by Zacknetic** [reply]
Cool, happy to help. 

In terms of that feature, once you have the time and motivation feel free to let me know here. You can join up as a contributor and the team and I would be happy to get you up to speed, answering any and all questions. 

Enjoy.

-Z

**Reply by EzzatQ** [reply]
Hey @Zacknetic 
So I ended up implementing the feature on a forked repository.
In the config.schema.json file, I added an option to add an array of uniqueId : controllerLogicType pairs which will override the automatically detected controllerLogicType.
Implemented the override code in platform.ts
You can check out my code here: [EzzatQ/HomebridgeMagicHome-DynamicPlatform
](https://github.com/EzzatQ/HomebridgeMagicHome-DynamicPlatform)

Still didn't create a pull request because I am not sure how to test the code.
the source code on github is in Typescript but the code I found in home bridge has 3 copies of each file: .d.ts, .d.ts.map and .js
I assume these are the files that were converted from the .ts files (I don't know anything about Typescript but managed to still write the code)

Could you please help me with that so I can test the code?

also I'd like to become a contributor, I have some other feature ideas I'd like to implement (like managing the saturation thresholds per device and not globally)

EQ

**Reply by Zacknetic** [reply]
Thanks for putting the effort in! This type of implementation is not consistent with the vision of the plugin which is why I was hoping you'd let me know before you started hard work on it. 

Join me on the [MagicHome discord server](https://discord.gg/c8xpmZSNZC) and give me @Zacknetic a shoutout when you arrive. We can discuss this further.
